### ISSUE TITLE
Decorators that change docstrings are not accounted for

### ISSUE THREAD

**Issue by blink1073** [davidhalter/jedi#117]
The following behavior was observed using the latest dev branch a86cfa2dd76b140e3c893beec03036921bfc3dfe.

The following code prints `old` instead of `new`, meaning that the decorator was not accounted for.

``` python
import jedi
code = """
def wrap(func):
    func.__doc__ = 'new'
    return func

@wrap
def orig():
    '''old'''
    pass

orig("""
script = jedi.Script(code, 11, 6, None)
print script.get_in_function_call().docstring()
```

Result:

``` python
'orig()\n\nold'
```

UPDATE: The following now works:

``` python
import jedi
code = """import functools; functools.partial("""
script = jedi.Script(code, 1, len(code), None)
print script.get_in_function_call().docstring()
```

Result:

``` python
'partial(func, *args, **keywords) - new function with partial application\nof the given arguments and keywords.'
```

**Comment by davidhalter** [comment]
Since I'm working on this kind of stuff now, I'm closing this in favor of #503. That issue is essentially the same than this one, but it is just about wraps, which is the remaining problem AFAIK.

Changing arbitrary docstrings is definitely not going to be supported in the future, however changing docstrings with wraps is going to be supported.

**Comment by blink1073** [comment]
Works for me, thanks @davidhalter!  ♥️

**Comment by davidhalter** [comment]
It's been a few years since you wrote this issue, haha. :)

**Comment by blink1073** [comment]
Any and all work you do here is much appreciated my friend

**Comment by jsh9** [comment]
Hello @davidhalter , I am encountering an issue with `jedi` (version 0.15.2).
This ticket here (https://github.com/spyder-ide/spyder/issues/11150) provides all the details.  Basically, Spyder relies on `jedi` to parse the docstring, and when a function is decorated with a decorator, this function's docstring cannot be correctly parsed.

**Comment by davidhalter** [comment]
Please use 0.16.0 and try again.

**Comment by jsh9** [comment]
I upgraded jedi to 0.16.0 (and parso is on 0.6.2), and still saw the same problem.  (I used Spyder to do the checking, because the example code above seemed outdated.)

**Comment by davidhalter** [comment]
The docstring I get is

```
FunctionType(*args: Any, **kwargs: Any) -> Any

A function

Parameters
----------

a : int
    Something.
```

This is probably expected. The signature seems wrong, but I'm not really sure it is, because there are `overload` definitions end in a callable, so that seems also correct. Can you give me more pointers on what's wrong according to you?

If you create a test script that helps a lot to reproduce. It also helps if you provide me with the actual output.

**Comment by davidhalter** [comment]
I'm realizing that there's something wrong with `Script.infer`. This is **not** the way how you should be getting docstrings. So if spyder does this, it's wrong. They should be using the `Script.help` function.

However I found the issue and fixed it.

**Comment by jsh9** [comment]
Hi @davidhalter :

Spyder was not able to extract the docstring of `func()`.  I relied on Spyder to do so, because I don't know how to extract docstrings outside of Spyder (if you can show me a way, I'd appreciate it).

Do still I need to ping Spyder's authors to fix the usage of `Spyder.infer`, or does your fix (https://github.com/davidhalter/jedi/commit/a2f4d1bbe7b95a06583a7588fd20bd1fd46a4572) solve the issue altogether?

Thanks!

**Comment by davidhalter** [comment]
> I relied on Spyder to do so, because I don't know how to extract docstrings outside of Spyder (if you can show me a way, I'd appreciate it).

It's as simple as:

```
import jedi
name, = jedi.Script(path='file.py').infer(line=2, column=3)
name.docstring()
```

Jedi's documentation is also available under https://jedi.readthedocs.org.

> Do still I need to ping Spyder's authors to fix the usage of Spyder.infer, or does your fix (a2f4d1b) solve the issue altogether?

No it's fine. They use a language server implementation that probably does quite a few things wrong. I'll do the talking at some point when Jedi 1.0 is out :).
